src/mapping/query/function/contains.rs

```diff
@@ -0,0 +1,169 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ContainsFn {
+    query: Box<dyn Function>,
+    substring: Box<dyn Function>,
+    case_sensitive: Option<Box<dyn Function>>,
+}
+
+impl ContainsFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        substring: &str,
+        case_sensitive: bool,
+    ) -> Self {
+        let substring = Box::new(Literal::from(Value::from(substring)));
+        let case_sensitive = Some(Box::new(Literal::from(Value::from(case_sensitive))) as _);
+
+        Self {
+            query,
+            substring,
+            case_sensitive,
+        }
+    }
+}
+
+impl Function for ContainsFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let substring = match self.substring.execute(ctx)? {
+            Value::Bytes(b) => b,
+            v => unexpected_type!(v),
+        };
```

Comment:
This is becoming a (very) common pattern in function implementations.

Might be worth creating a small macro to do something like:

```rust
let substring: bytes::Bytes = required!(self.substring, Value::Bytes(b) => b, Value::Integer(i) => { ... });
let case_sensitive: Option<bool> = optional!(self.case_sensitive, Value::Boolean(v) => v);
```

_(the type signature is only there for documentation purposes)_

cc @FungusHumungus

Reply:
I like this idea a lot!

Reply:
What about a set of methods on `Value` that work similar to `.unwrap`? It does mean one or two method per value type, but we do have precedent for using a macro to do that kind of code creation (ref `impl_valuekind_from_integer`).
```rust
    let substring = self.substring.execute(ctx)?.expect_bytes();
```
This gets awkward for options, but I think there are possibilities there too with something like `self.case_sensitive.map_or(…)`.

One question to answer here would be: what form will be easier to understand and modify later?

Reply:
That's an interesting idea @bruceg. One issue with that, is that there are many parameters that accept more than one value variant, and while we could model that on `Value` as well, I'm not so sure it's valuable outside the work we're doing in `remap`.

Reply:
I added `optional!` and `required!` in ddf3ece.

We can iterate on it going forward if we come up with better solutions, but I think this is already an improvement. I will apply the changes to the other remap functions in a separate PR.